File: crates/bodhi/src/test-utils/msw-v2/handlers/toolsets.ts
/**
 * MSW v2 handlers for toolsets endpoints - instance-based architecture
 */
import { http, HttpResponse } from 'msw';

import { BODHI_API_BASE } from '@/hooks/useQuery';
import type { ToolsetResponse, ToolsetTypeResponse, AppToolsetConfigResponse } from '@/hooks/useToolsets';

// ============================================================================
// Mock Data
// ============================================================================

export const mockToolset: ToolsetResponse = {
  id: 'uuid-test-toolset',
  name: 'my-exa-search',
  scope_uuid: '4ff0e163-36fb-47d6-a5ef-26e396f067d6',
  scope: 'scope_toolset-builtin-exa-web-search',
  description: 'Test toolset',
  enabled: true,
  has_api_key: true,
  tools: [
    {
      type: 'function',
      function: {
        name: 'search',
        description: 'Search the web using Exa AI',
        parameters: {
          type: 'object',
          properties: {
            query: { type: 'string', description: 'The search query' },
          },
          required: ['query'],
        },
      },
    },
  ],
  created_at: '2024-01-01T00:00:00Z',
  updated_at: '2024-01-01T00:00:00Z',
};

export const mockToolsetTypes = [
  {
    scope: 'scope_toolset-builtin-exa-web-search',
    scope_uuid: '4ff0e163-36fb-47d6-a5ef-26e396f067d6',
    enabled: true,
    updated_by: 'system',
    created_at: '2024-01-01T00:00:00Z',
    updated_at: '2024-01-01T00:00:00Z',
  },
];

export const mockType: ToolsetTypeResponse = {
  scope_uuid: '4ff0e163-36fb-47d6-a5ef-26e396f067d6',
  scope: 'scope_toolset-builtin-exa-web-search',
  name: 'Exa Web Search',
  description: 'Search the web using Exa AI',
  app_enabled: true,
  tools: [
    {
      type: 'function',
      function: {
        name: 'search',
        description: 'Search the web using Exa AI',
        parameters: {
          type: 'object',
          properties: {
            query: { type: 'string', description: 'The search query' },
          },
          required: ['query'],
        },
      },
    },
  ],
};

// ============================================================================
// Handler Factories
// ============================================================================

/**
 * Mock GET /bodhi/v1/toolsets - List user's toolsets
 */
export function mockListToolsets(toolsets: ToolsetResponse[] = [mockToolset], toolset_types = mockToolsetTypes) {
  return http.get(`${BODHI_API_BASE}/toolsets`, () => HttpResponse.json({ toolsets, toolset_types }));
}

/**
 * Mock GET /bodhi/v1/toolsets/:id - Get single toolset
 */
export function mockGetToolset(toolset: ToolsetResponse = mockToolset) {
  return http.get(`${BODHI_API_BASE}/toolsets/:id`, () => HttpResponse.json(toolset));
}

/**
 * Mock POST /bodhi/v1/toolsets - Create new toolset
 */
export function mockCreateToolset(response: ToolsetResponse = mockToolset) {
  return http.post(`${BODHI_API_BASE}/toolsets`, () => HttpResponse.json(response, { status: 201 }));
}

/**
 * Mock PUT /bodhi/v1/toolsets/:id - Update toolset
 */
export function mockUpdateToolset(response: ToolsetResponse = mockToolset) {
  return http.put(`${BODHI_API_BASE}/toolsets/:id`, () => HttpResponse.json(response));
}

/**
 * Mock DELETE /bodhi/v1/toolsets/:id - Delete toolset
 */
export function mockDeleteToolset() {
  return http.delete(`${BODHI_API_BASE}/toolsets/:id`, () => new HttpResponse(null, { status: 204 }));
}

/**
 * Mock GET /bodhi/v1/toolset_types - List toolset types
 */
export function mockListTypes(types: ToolsetTypeResponse[] = [mockType]) {
  return http.get(`${BODHI_API_BASE}/toolset_types`, () => HttpResponse.json({ types }));
}

/**
 * Mock PUT /bodhi/v1/toolset_types/:scope/app-config - Enable toolset type (admin)
 */
export function mockEnableType(response?: AppToolsetConfigResponse) {
  const defaultResponse: AppToolsetConfigResponse = {
    scope: 'scope_toolset-builtin-exa-web-search',
    scope_uuid: '4ff0e163-36fb-47d6-a5ef-26e396f067d6',
    enabled: true,
    updated_by: 'admin123',
    created_at: '2024-01-01T00:00:00Z',
    updated_at: '2024-01-01T00:00:00Z',
  };
  return http.put(`${BODHI_API_BASE}/toolset_types/:scope/app-config`, () =>
    HttpResponse.json(response || defaultResponse)
  );
}

/**
 * Mock DELETE /bodhi/v1/toolset_types/:scope/app-config - Disable toolset type (admin)
 */
export function mockDisableType(response?: AppToolsetConfigResponse) {
  const defaultResponse: AppToolsetConfigResponse = {
    scope: 'scope_toolset-builtin-exa-web-search',
    scope_uuid: '4ff0e163-36fb-47d6-a5ef-26e396f067d6',
    enabled: false,
    updated_by: 'admin123',
    created_at: '2024-01-01T00:00:00Z',
    updated_at: '2024-01-01T00:00:00Z',
  };
  return http.delete(`${BODHI_API_BASE}/toolset_types/:scope/app-config`, () =>
    HttpResponse.json(response || defaultResponse)
  );
}

// ============================================================================
// Error Handlers
// ============================================================================

/**
 * Mock error for toolsets list endpoint
 */
export function mockListToolsetsError({
  message = 'Failed to fetch toolsets',
  code = 'internal_server_error',
  type = 'internal_server_error',
  status = 500,
}: { message?: string; code?: string; type?: string; status?: number } = {}) {
  return http.get(`${BODHI_API_BASE}/toolsets`, () =>
    HttpResponse.json({ error: { message, code, type } }, { status })
  );
}

/**
 * Mock error for get toolset endpoint
 */
export function mockGetToolsetError({
  message = 'Toolset not found',
  code = 'not_found',
  type = 'not_found_error',
  status = 404,
}: { message?: string; code?: string; type?: string; status?: number } = {}) {
  return http.get(`${BODHI_API_BASE}/toolsets/:id`, () =>
    HttpResponse.json({ error: { message, code, type } }, { status })
  );
}

/**
 * Mock error for create toolset endpoint
 */
export function mockCreateToolsetError({
  message = 'Failed to create toolset',
  code = 'internal_server_error',
  type = 'internal_server_error',
  status = 500,
}: { message?: string; code?: string; type?: string; status?: number } = {}) {
  return http.post(`${BODHI_API_BASE}/toolsets`, () =>
    HttpResponse.json({ error: { message, code, type } }, { status })
  );
}

/**
 * Mock error for update toolset endpoint
 */
export function mockUpdateToolsetError({
  message = 'Failed to update toolset',
  code = 'internal_server_error',
  type = 'internal_server_error',
  status = 500,
}: { message?: string; code?: string; type?: string; status?: number } = {}) {
  return http.put(`${BODHI_API_BASE}/toolsets/:id`, () =>
    HttpResponse.json({ error: { message, code, type } }, { status })
  );
}

// ============================================================================
// Default Handlers
// ============================================================================

export const toolsetsHandlers = [
  mockListToolsets(),
  mockGetToolset(),
  mockCreateToolset(),
  mockUpdateToolset(),
  mockDeleteToolset(),
  mockListTypes(),
  mockEnableType(),
  mockDisableType(),
];

File: crates/routes_app/src/routes_toolsets/toolsets.rs
use crate::{
  ApiKeyUpdateDto, CreateToolsetRequest, ExecuteToolsetRequest, ListToolsetTypesResponse,
  ListToolsetsResponse, ToolsetResponse, ToolsetValidationError, UpdateToolsetRequest,
  ENDPOINT_TOOLSETS, ENDPOINT_TOOLSET_TYPES,
};
use auth_middleware::ExtractUserId;
use axum::{
  extract::{Path, State},
  http::StatusCode,
  routing::{delete, get, post, put},
  Json, Router,
};
use objs::API_TAG_TOOLSETS;
use objs::{ApiError, Toolset, ToolsetExecutionResponse};
use server_core::RouterState;
use services::db::ApiKeyUpdate;
use std::sync::Arc;
use validator::Validate;

// ============================================================================
// Router Configuration
// ============================================================================

pub fn routes_toolsets(state: Arc<dyn RouterState>) -> Router {
  Router::new()
    // Toolset CRUD
    .route("/toolsets", get(list_toolsets_handler))
    .route("/toolsets", post(create_toolset_handler))
    .route("/toolsets/{id}", get(get_toolset_handler))
    .route("/toolsets/{id}", put(update_toolset_handler))
    .route("/toolsets/{id}", delete(delete_toolset_handler))
    // Execute (middleware at routes level)
    .route(
      "/toolsets/{id}/execute/{method}",
      post(execute_toolset_handler),
    )
    // Type listing (separate namespace avoids {id} collision)
    .route("/toolset_types", get(list_toolset_types_handler))
    // Admin enable/disable
    .route(
      "/toolset_types/{toolset_type}/app-config",
      put(enable_type_handler),
    )
    .route(
      "/toolset_types/{toolset_type}/app-config",
      delete(disable_type_handler),
    )
    .with_state(state)
}

// ============================================================================
// Toolset CRUD Handlers
// ============================================================================

/// List all toolsets for the authenticated user
#[utoipa::path(
  get,
  path = ENDPOINT_TOOLSETS,
  tag = API_TAG_TOOLSETS,
  operation_id = "listToolsets",
  responses(
    (status = 200, description = "List of user's toolsets", body = ListToolsetsResponse),
  ),
  security(("bearer" = []))
)]
pub async fn list_toolsets_handler(
  ExtractUserId(user_id): ExtractUserId,
  State(state): State<Arc<dyn RouterState>>,
) -> Result<Json<ListToolsetsResponse>, ApiError> {
  let tool_service = state.app_service().tool_service();

  let toolsets = tool_service.list(&user_id).await?;

  // Enrich each toolset with type information
  let mut responses = Vec::new();
  for toolset in toolsets {
    responses.push(toolset_to_response(toolset, &tool_service).await?);
  }

  // Populate toolset_types from database
  let toolset_types = tool_service.list_app_toolset_configs().await?;

  Ok(Json(ListToolsetsResponse {
    toolsets: responses,
    toolset_types,
  }))
}

/// Create a new toolset
#[utoipa::path(
  post,
  path = ENDPOINT_TOOLSETS,
  tag = API_TAG_TOOLSETS,
  operation_id = "createToolset",
  request_body = CreateToolsetRequest,
  responses(
    (status = 201, description = "Toolset created", body = ToolsetResponse),
    (status = 400, description = "Validation error"),
    (status = 409, description = "Name already exists"),
  ),
  security(("bearer" = []))
)]
pub async fn create_toolset_handler(
  ExtractUserId(user_id): ExtractUserId,
  State(state): State<Arc<dyn RouterState>>,
  Json(request): Json<CreateToolsetRequest>,
) -> Result<(StatusCode, Json<ToolsetResponse>), ApiError> {
  request
    .validate()
    .map_err(|e| ToolsetValidationError::Validation(e.to_string()))?;
  let tool_service = state.app_service().tool_service();

  let toolset = tool_service
    .create(
      &user_id,
      &request.toolset_type,
      &request.slug,
      request.description,
      request.enabled,
      request.api_key,
    )
    .await?;

  let response = toolset_to_response(toolset, &tool_service).await?;
  Ok((StatusCode::CREATED, Json(response)))
}

/// Get a specific toolset by ID
#[utoipa::path(
  get,
  path = ENDPOINT_TOOLSETS.to_owned() + "/{id}",
  tag = API_TAG_TOOLSETS,
  operation_id = "getToolset",
  params(
    ("id" = String, Path, description = "Toolset instance UUID")
  ),
  responses(
    (status = 200, description = "Toolset", body = ToolsetResponse),
    (status = 404, description = "Toolset not found or not owned"),
  ),
  security(("bearer" = []))
)]
pub async fn get_toolset_handler(
  ExtractUserId(user_id): ExtractUserId,
  State(state): State<Arc<dyn RouterState>>,
  Path(id): Path<String>,
) -> Result<Json<ToolsetResponse>, ApiError> {
  let tool_service = state.app_service().tool_service();

  let toolset = tool_service
    .get(&user_id, &id)
    .await?
    .ok_or_else(|| objs::EntityError::NotFound("Toolset".to_string()))?;

  let response = toolset_to_response(toolset, &tool_service).await?;
  Ok(Json(response))
}

/// Update a toolset (full PUT semantics)
#[utoipa::path(
  put,
  path = ENDPOINT_TOOLSETS.to_owned() + "/{id}",
  tag = API_TAG_TOOLSETS,
  operation_id = "updateToolset",
  params(
    ("id" = String, Path, description = "Toolset instance UUID")
  ),
  request_body = UpdateToolsetRequest,
  responses(
    (status = 200, description = "Toolset updated", body = ToolsetResponse),
    (status = 400, description = "Validation error"),
    (status = 404, description = "Toolset not found or not owned"),
    (status = 409, description = "Name already exists"),
  ),
  security(("bearer" = []))
)]
pub async fn update_toolset_handler(
  ExtractUserId(user_id): ExtractUserId,
  State(state): State<Arc<dyn RouterState>>,
  Path(id): Path<String>,
  Json(request): Json<UpdateToolsetRequest>,
) -> Result<Json<ToolsetResponse>, ApiError> {
  request
    .validate()
    .map_err(|e| ToolsetValidationError::Validation(e.to_string()))?;
  let tool_service = state.app_service().tool_service();

  let api_key_update = match request.api_key {
    ApiKeyUpdateDto::Keep => ApiKeyUpdate::Keep,
    ApiKeyUpdateDto::Set(value) => ApiKeyUpdate::Set(value),
  };

  let toolset = tool_service
    .update(
      &user_id,
      &id,
      &request.slug,
      request.description,
      request.enabled,
      api_key_update,
    )
    .await?;

  let response = toolset_to_response(toolset, &tool_service).await?;
  Ok(Json(response))
}

/// Delete a toolset
#[utoipa::path(
  delete,
  path = ENDPOINT_TOOLSETS.to_owned() + "/{id}",
  tag = API_TAG_TOOLSETS,
  operation_id = "deleteToolset",
  params(
    ("id" = String, Path, description = "Toolset instance UUID")
  ),
  responses(
    (status = 204, description = "Toolset deleted"),
    (status = 404, description = "Toolset not found or not owned"),
  ),
  security(("bearer" = []))
)]
pub async fn delete_toolset_handler(
  ExtractUserId(user_id): ExtractUserId,
  State(state): State<Arc<dyn RouterState>>,
  Path(id): Path<String>,
) -> Result<StatusCode, ApiError> {
  let tool_service = state.app_service().tool_service();

  tool_service.delete(&user_id, &id).await?;

  Ok(StatusCode::NO_CONTENT)
}

// ============================================================================
// Execute Handler
// ============================================================================

/// Execute a tool method on a toolset
#[utoipa::path(
  post,
  path = ENDPOINT_TOOLSETS.to_owned() + "/{id}/execute/{method}",
  tag = API_TAG_TOOLSETS,
  operation_id = "executeToolset",
  params(
    ("id" = String, Path, description = "Toolset instance UUID"),
    ("method" = String, Path, description = "Tool method name")
  ),
  request_body = ExecuteToolsetRequest,
  responses(
    (status = 200, description = "Tool execution result", body = ToolsetExecutionResponse),
    (status = 400, description = "Validation error or toolset not configured"),
    (status = 404, description = "Toolset or method not found"),
  ),
  security(("bearer" = []))
)]
pub async fn execute_toolset_handler(
  ExtractUserId(user_id): ExtractUserId,
  State(state): State<Arc<dyn RouterState>>,
  Path((id, method)): Path<(String, String)>,
  Json(request): Json<ExecuteToolsetRequest>,
) -> Result<Json<ToolsetExecutionResponse>, ApiError> {
  let tool_service = state.app_service().tool_service();

  let response = tool_service
    .execute(&user_id, &id, &method, request.into())
    .await?;

  Ok(Json(response))
}

// ============================================================================
// Toolset Type Handlers (Admin)
// ============================================================================

/// List all available toolset types with their tools
#[utoipa::path(
  get,
  path = ENDPOINT_TOOLSET_TYPES,
  tag = API_TAG_TOOLSETS,
  operation_id = "listToolsetTypes",
  responses(
    (status = 200, description = "List of all toolset types", body = ListToolsetTypesResponse),
  ),
  security(("bearer" = []))
)]
pub async fn list_toolset_types_handler(
  State(state): State<Arc<dyn RouterState>>,
) -> Result<Json<ListToolsetTypesResponse>, ApiError> {
  let tool_service = state.app_service().tool_service();
  let types = tool_service.list_types();
  Ok(Json(ListToolsetTypesResponse { types }))
}

/// Enable a toolset type at app level (Admin only)
#[utoipa::path(
  put,
  path = "/bodhi/v1/toolset_types/{toolset_type}/app-config",
  tag = API_TAG_TOOLSETS,
  operation_id = "enableToolsetType",
  params(
    ("toolset_type" = String, Path, description = "Toolset type identifier (e.g., 'builtin-exa-search')")
  ),
  responses(
    (status = 200, description = "Toolset type enabled", body = objs::AppToolsetConfig),
    (status = 404, description = "Toolset type not found"),
  ),
  security(("bearer" = []))
)]
pub async fn enable_type_handler(
  ExtractUserId(updated_by): ExtractUserId,
  State(state): State<Arc<dyn RouterState>>,
  Path(toolset_type): Path<String>,
) -> Result<Json<objs::AppToolsetConfig>, ApiError> {
  let tool_service = state.app_service().tool_service();

  let config = tool_service
    .set_app_toolset_enabled(&toolset_type, true, &updated_by)
    .await?;

  Ok(Json(config))
}

/// Disable a toolset type at app level (Admin only)
#[utoipa::path(
  delete,
  path = "/bodhi/v1/toolset_types/{toolset_type}/app-config",
  tag = API_TAG_TOOLSETS,
  operation_id = "disableToolsetType",
  params(
    ("toolset_type" = String, Path, description = "Toolset type identifier (e.g., 'builtin-exa-search')")
  ),
  responses(
    (status = 200, description = "Toolset type disabled", body = objs::AppToolsetConfig),
    (status = 404, description = "Toolset type not found"),
  ),
  security(("bearer" = []))
)]
pub async fn disable_type_handler(
  ExtractUserId(updated_by): ExtractUserId,
  State(state): State<Arc<dyn RouterState>>,
  Path(toolset_type): Path<String>,
) -> Result<Json<objs::AppToolsetConfig>, ApiError> {
  let tool_service = state.app_service().tool_service();

  let config = tool_service
    .set_app_toolset_enabled(&toolset_type, false, &updated_by)
    .await?;

  Ok(Json(config))
}

// ============================================================================
// Conversion Helpers
// ============================================================================

async fn toolset_to_response(
  toolset: Toolset,
  tool_service: &Arc<dyn services::ToolService>,
) -> Result<ToolsetResponse, ApiError> {
  // Get type information for enrichment
  let type_def = tool_service
    .get_type(&toolset.toolset_type)
    .ok_or_else(|| services::ToolsetError::InvalidToolsetType(toolset.toolset_type.clone()))?;

  Ok(ToolsetResponse {
    id: toolset.id,
    slug: toolset.slug,
    toolset_type: toolset.toolset_type,
    description: toolset.description,
    enabled: toolset.enabled,
    has_api_key: toolset.has_api_key,
    tools: type_def.tools,
    created_at: toolset.created_at,
    updated_at: toolset.updated_at,
  })
}
